### PR TITLE
[RISCV] Teach RISCVMakeCompressible handle Zca/Zcf/Zce/Zcd.

### PR DESCRIPTION
**Pull request by yetingk**
Make targets which don't have C but have Zca/Zcf/Zce/Zcd benefit from this pass.

### CODE REVIEW

**Comment by asb** [comment]
Thanks - I've left a couple of comments inline. The logic would be simplified if we moved to C implying the appropriate Zc* extensions as you wouldn't need to check for C || Zc*. But that's a separate topic and we held back due to concern about interacting with older assemblers. It might merit a TODO note though.

Are you planning to follow-up with a patch to teach isCompressibleLoad and isCompressibleStore about half/byte loads+stores where Zcb is available?

File: llvm/lib/Target/RISCV/RISCVSubtarget.h [hunk 1]
```diff
@@ -143,6 +143,10 @@ class RISCVSubtarget : public RISCVGenSubtargetInfo {
 #include "RISCVGenSubtargetInfo.inc"
 
   bool hasStdExtCOrZca() const { return HasStdExtC || HasStdExtZca; }
+  bool hasStdExtCOrZcd() const { return HasStdExtC || HasStdExtZcd; }
+  bool hasStdExtCOrZcfOrZce() const {
```
[the remark this answers is not in the document]
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVSubtarget.h:147]: We do have a `HasStdExtCOrZcfOrZce` as a tablegen predicate already. There was a reason I included in Zce in there, but I don't remember exactly why.

File: llvm/lib/Target/RISCV/RISCVMakeCompressible.cpp [hunk 2]
```diff
@@ -324,8 +340,7 @@ bool RISCVMakeCompressibleOpt::runOnMachineFunction(MachineFunction &Fn) {
   const RISCVInstrInfo &TII = *STI.getInstrInfo();
 
   // This optimization only makes sense if compressed instructions are emitted.
-  // FIXME: Support Zca, Zcf, Zcd granularity.
-  if (!STI.hasStdExtC())
+  if (!STI.hasStdExtZca() && !STI.hasStdExtCOrZcfOrZce() && !STI.hasStdExtZcd())
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVMakeCompressible.cpp:343]: Should be enough to check `STI.hasStdExtCOrZca();` here I think?

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**